### PR TITLE
576 get cmd class factories

### PR DESCRIPTION
**Pull request by evrose54**
# Reviewer Checklist

* [ ] Required ***existing tests*** pass (ie full_test.sh, others as appropriate)
* [ ] Required ***unit tests*** modified for updated functionality
* [ ] NO REQUIRED ***integration tests*** (explain why not required)
* [ ] Required ***documentation*** added for new/modified functionality
* [ ] Required ***release notes*** added for new/modified functionality
* [ ] NO REQUIRED ***updates to other repos*** (explain why not required)

# Related Issues
fixes NRLMMD-GEOIPS/geoips#573
fixes NRLMMD-GEOIPS/geoips#576

# Testing Instructions
Run ``pytest -v ./tests/unit_tests/commandline/``

# Summary
This branch refactors commands which use the form of ``geoips get <interface_name>`` to
make use of class factories. Before, we had to run commands such as:

* ``geoips get interface <interface_name>``
* ``geoips get family <interface_name> <family_name>``
* ``geoips get plugin <interface_name> <family_name>``

The aforementioned commands were overly verbose and needed some refactoring to reduce
the levels required to execute a certain command. In this branch, we refactored
``GeoipsGetInterface`` to be a base command for ``get <interface_name>``,
``get <interface_name> family <family_name>``, and
``get <interface_name> <plugin_name>``. This resulted in the removal of
``GeoipsGetFamily`` and ``GeoipsGetPlugin`` and the refactoring of
``GeoipsGetInterface`` to include the functionality of those other commands. We can use
class factories for each interface to execute these commands now.

On top of this, we implemented aliases for a variety of CLI commands. For an exact
listing of aliases available for each command, see
``geoips/commandline/ancillary_info/alias_mapping.yaml``, which includes aliases
supported for each command. Now, we can run a verbose command such as:

* ``geoips list algorithms -p geoips``

and replace it with

* ``geoips ls algs -p geoips``

This also reduces the verbosity of certain CLI commands. Note, the unit tests have been
modified to reflect these changes.

    modified: .gitignore
    modified: pyproject.toml
    modified: docs/source/_templates/conf_PKG.py
    modified: docs/source/userguide/command_line.rst
    modified: geoips/commandline/ancillary_info/cmd_instructions.yaml
    modified: geoips/commandline/cmd_instructions.py
    modified: geoips/commandline/commandline_interface.py
    modified: geoips/commandline/geoips_command.py
    modified: geoips/commandline/geoips_get.py
    modified: tests/unit_tests/commandline/cli_top_level_tester.py
    modified: tests/unit_tests/commandline/test_geoips_get_family.py
    modified: tests/unit_tests/commandline/test_geoips_get_interface.py
    modified: tests/unit_tests/commandline/test_geoips_get_package.py
    modified: tests/unit_tests/commandline/test_geoips_get_plugin.py
    modified: tests/unit_tests/commandline/test_geoips_list_interface.py
    modified: tests/unit_tests/commandline/test_geoips_list_interfaces.py
    modified: tests/unit_tests/commandline/test_geoips_list_packages.py
    modified: tests/unit_tests/commandline/test_geoips_list_plugins.py
    modified: tests/unit_tests/commandline/test_geoips_list_scripts.py
    modified: tests/unit_tests/commandline/test_geoips_list_test_datasets.py
    modified: tests/unit_tests/commandline/test_geoips_list_unit_tests.py
    modified: tests/unit_tests/commandline/test_get_commandline_instructions.py
    added: geoips/commandline/ancillary_info/alias_mapping.yaml

### CODE REVIEW

File: geoips/commandline/cmd_instructions.py
Comment (jsolbrig) [inline geoips/commandline/cmd_instructions.py:54]: I would prefer that `fname` takes a full filename including the extension. I think we should use `.yaml` here since that is what is normally referenced. The `.json` files are just an under-the-hood efficiency update.

File: geoips/commandline/cmd_instructions.py
Comment (jsolbrig) [inline geoips/commandline/cmd_instructions.py:23]: See this comment: https://github.com/NRLMMD-GEOIPS/geoips/pull/644/files#r1644854681

File: geoips/commandline/ancillary_info/cmd_instructions.yaml
Comment (jsolbrig) [inline geoips/commandline/ancillary_info/cmd_instructions.yaml:24]: Grammar:
> Retrieve information about a GeoIPS artifact.

File: geoips/commandline/cmd_instructions.py
Comment (evrose54) [reply geoips/commandline/cmd_instructions.py]: Changed this to a full name as in ``cmd_instructions.yaml``

File: geoips/commandline/ancillary_info/cmd_instructions.yaml
Comment (jsolbrig) [reply geoips/commandline/ancillary_info/cmd_instructions.yaml:24]: I think we should also consider how this can be made more specific to each sub-command so that, for example, `geoips get algorithm -h` can show something specific to algorithms.

Comment (evrose54) [reply geoips/commandline/ancillary_info/cmd_instructions.yaml:24]: Fixed!

Comment (evrose54) [reply geoips/commandline/ancillary_info/cmd_instructions.yaml]: I'll make an issue for making the help information more specific about the command ran. I think that's going to require a decent amount of testing and refactoring and should likely be put in a different PR.

Comment (evrose54) [reply geoips/commandline/ancillary_info/cmd_instructions.yaml]: See https://github.com/NRLMMD-GEOIPS/geoips/issues/645

File: geoips/commandline/cmd_instructions.py
Comment (evrose54) [reply geoips/commandline/cmd_instructions.py]: Fixed

**Review by jsolbrig — APPROVED** [review]

File: geoips/commandline/geoips_list.py
Comment (mindyls) [inline geoips/commandline/geoips_list.py:666]: Is this the loop you were going to move outside the class in another PR?

File: tests/unit_tests/commandline/cmd_instructions/invalid_json/cmd_instructions.json
Comment (mindyls) [inline tests/unit_tests/commandline/cmd_instructions/invalid_json/cmd_instructions.json:3]: I have noticed this file gets modified sometimes - was this supposed to be checked in?

File: tests/unit_tests/commandline/cmd_instructions/json_newer/cmd_instructions.json
Comment (mindyls) [inline tests/unit_tests/commandline/cmd_instructions/json_newer/cmd_instructions.json:2]: Should this have been checked in?

File: tests/unit_tests/commandline/test_geoips_get_package.py
Comment (mindyls) [inline tests/unit_tests/commandline/test_geoips_get_package.py:25]: These aliases are not included in the alias mapping? Or this is explicitly being set for tests?

**Review by mindyls — DISMISSED** [review]

File: geoips/commandline/geoips_get.py
Comment (evrose54) [reply geoips/commandline/geoips_get.py:24]: Yes, since that's largely just a nomenclature change, I'm waiting to implement that until we get all of the CLI PRs merged in. Don't want to deal with annoying conflicts for a non-functional change.

File: geoips/commandline/geoips_list.py
Comment (evrose54) [reply geoips/commandline/geoips_list.py:666]: This loop is the 'moved loop'. Originally this loop was implemented in ``CommandClassFactory``, however we wanted that class to only generate one class per instantiation. This is where we'll want to keep this loop for the time being, unless we think of a different way to implement this.

File: tests/unit_tests/commandline/cmd_instructions/invalid_json/cmd_instructions.json
Comment (evrose54) [reply tests/unit_tests/commandline/cmd_instructions/invalid_json/cmd_instructions.json:3]: It is supposed to be checked in. This is used for the unit tests. I'm happy just removing the test cases under ``tests/unit_tests/commandline/cmd_instructions`` as they make testing more complex and cause problems with git. Up to you on this one though.

File: tests/unit_tests/commandline/cmd_instructions/json_newer/cmd_instructions.json
Comment (evrose54) [reply tests/unit_tests/commandline/cmd_instructions/json_newer/cmd_instructions.json:2]: See comment above.

File: tests/unit_tests/commandline/test_geoips_get_package.py
Comment (evrose54) [reply tests/unit_tests/commandline/test_geoips_get_package.py]: These are included in alias mapping. But are listed explicitly here. I've changed the code to make this dynamic.

File: geoips/commandline/ancillary_info/cmd_instructions.yaml
Comment (evrose54) [reply geoips/commandline/ancillary_info/cmd_instructions.yaml:134]: I'm leaning towards keeping it until we've made a final decision on what the help messages will contain, a list of stable commands, etc. Better to have it and not use it rather than making code changes and having to rewrite it. That being said, I'm happy to remove it if that's what we want to do.

File: geoips/commandline/geoips_list.py
Comment (jsolbrig) [reply geoips/commandline/geoips_list.py:678]: See #708 for the future of this.

File: docs/source/releases/v1_13_0a0.rst
Comment (jsolbrig) [inline docs/source/releases/v1_13_0a0.rst:1]: These changes should be moved into the `docs/source/releases/latest/` directory.

**Review by jsolbrig — APPROVED** [review]

**Review by jsolbrig — APPROVED** [review]

**Review by jsolbrig — APPROVED** [review]

**Review by jsolbrig — APPROVED** [review]

File: geoips/commandline/geoips_get.py
Comment (mindyls) [reply geoips/commandline/geoips_get.py:27]: Good plan ...

File: geoips/commandline/geoips_get.py
Comment (mindyls) [inline geoips/commandline/geoips_get.py:40]: Did two lines get combined here?

**Review by mindyls — CHANGES_REQUESTED** [review]
I don't know how to get conversations to show up instead of review comments on my phone... so requesting changes so my conversations will show up ...

**Review by mindyls — DISMISSED** [review]
If the build docs is resolved this looks good to me.  Note I have not tested command line,  so assuming functionality works as expected.  I will test with future PR.

File: geoips/commandline/geoips_get.py
Comment (evrose54) [reply geoips/commandline/geoips_get.py]: Good catch. Removing that now.

File: docs/source/releases/latest/576-get-cmd-class-factories.yaml
Comment (evrose54) [reply docs/source/releases/latest/576-get-cmd-class-factories.yaml:3]: This isn't an ``index.rst`` problem. Seems like the CI is not installing the new dependency I added. The documentation builds fine for me locally. Probably an issue that we should look into.

**Review by jsolbrig — APPROVED** [review]

**Review by mindyls — APPROVED** [review]
Looks good

**Review by jsolbrig — APPROVED** [review]

**Pull request merged**
